searx/engines/command.py

```diff
@@ -0,0 +1,182 @@
+'''
+searx is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+searx is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with searx. If not, see < http://www.gnu.org/licenses/ >.
+'''
+
+
+from os.path import expanduser, isabs, realpath, commonprefix
+from re import MULTILINE, search as re_search
+from shlex import split as shlex_split
+from subprocess import Popen, PIPE
+from time import time
+from threading import Thread
+
+from searx import logger
+from searx.utils import str_split
+
+
+offline = True
+paging = True
+command = []
+delimiter = {}
+parse_regex = {}
+query_type = ''
+query_enum = []
+working_dir = realpath('.')
+result_separator = '\n'
+result_template = 'key-value.html'
+timeout = 4.0
+
+_command_logger = logger.getChild('command')
+
+
+def init(engine_settings):
+    check_parsing_options(engine_settings)
+
+    if 'command' not in engine_settings:
+        raise ValueError('engine command : missing configuration key: command')
+
+    global command, working_dir, result_template, delimiter, parse_regex, timeout
+
+    command = engine_settings['command']
+
+    if 'working_dir' in engine_settings:
+        working_dir = engine_settings['working_dir']
+        if not isabs(engine_settings['working_dir']):
+            working_dir = realpath(working_dir)
+
+    if 'parse_regex' in engine_settings:
+        parse_regex = engine_settings['parse_regex']
+    if 'delimiter' in engine_settings:
+        delimiter = engine_settings['delimiter']
+
+
+def search(query, params):
+    cmd = _get_command_to_run(query)
+    if not cmd:
+        return []
+
+    results = []
+    process = Popen(cmd, stdout=PIPE, stderr=PIPE)
+    reader_thread = Thread(target=_get_results_from_process, args=(results, process, params['pageno']))
+    reader_thread.start()
+
+    return_code = process.wait(timeout=timeout)
```

Comment:
if the wait fail, the process may be still alive.

I would change to something like:
```
process = Popen(...)
try:
        reader_thread = Thread(...)
        try:
                reader_thread.start()
                return_code = process.wait(timeout=timeout)
                if return_code != 0:
                        raise ...
        finally:
                reader_thread.join(timeout=timeout)
                # more tricky: there is no clean to kill a process
finally:
        if process.is_alive():
            process.kill()
```